cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,445 @@
+        ):
+            raise ValueError("Input circuit must only have ops that act on 1 or 2 qubits.")
+
+        # 1. Do the initial mapping of logical to physical qubits.
+        if initial_mapper is None:
+            initial_mapper = line_initial_mapper.LineInitialMapper(self.device_graph)
+        initial_mapping = initial_mapper.initial_mapping(circuit)
+
+        # 2. Construct a mapping manager that implicitly keeps track of this mapping and provides
+        # convinience methods over the image of the map on the device graph.
+        mm = mapping_manager.MappingManager(self.device_graph, initial_mapping)
+
+        # 3. Get two_qubit_ops and single-qubit operations.
+        two_qubit_ops, single_qubit_ops = self._get_two_qubit_ops_and_single_qubit_ops(circuit)
+
+        # 4. Do the routing and save the routed circuit as a list of moments.
+        routed_ops = self._route(
+            mm,
+            two_qubit_ops,
+            single_qubit_ops,
+            max_search_radius,
+            tag_inserted_swaps=tag_inserted_swaps,
+        )
+
+        # 5. Return the routed circuit by packing each inner list of ops as densely as posslbe and
+        # preserving outer moment structure. Also return initial map and swap permutation map.
+        return (
+            circuits.Circuit(circuits.Circuit(m) for m in routed_ops),
+            initial_mapping,
+            {initial_mapping[k]: v for k, v in mm.map.items()},
+        )
+
+    @classmethod
+    def _get_two_qubit_ops_and_single_qubit_ops(
+        cls, circuit: 'cirq.AbstractCircuit'
+    ) -> Tuple[List[List['cirq.Operation']], List[List['cirq.Operation']]]:
+        """Gets the single and two qubit operations of the circuit.
+
+        By construction, single-qubit operations are inserted after two-qubit operations when
+        reinserting in the routed circuit.
+        """
+        two_qubit_circuit = circuits.Circuit()
+        single_qubit_ops: List[List[cirq.Operation]] = []
+        for moment in circuit:
+            for op in moment:
+                timestep = two_qubit_circuit.earliest_available_moment(op)
+                single_qubit_ops.extend([] for _ in range(timestep + 1 - len(single_qubit_ops)))
+                two_qubit_circuit.append(
+                    circuits.Moment() for _ in range(timestep + 1 - len(two_qubit_circuit))
+                )
+                if protocols.num_qubits(op) == 2 and not protocols.is_measurement(op):
+                    two_qubit_circuit[timestep] = two_qubit_circuit[timestep].with_operation(op)
+                else:
+                    single_qubit_ops[timestep].append(op)
+        two_qubit_ops = [list(m) for m in two_qubit_circuit]
+        return two_qubit_ops, single_qubit_ops
+
+    @classmethod
+    def _route(
+        cls,
+        mm: mapping_manager.MappingManager,
+        two_qubit_ops: List[List['cirq.Operation']],
+        single_qubit_ops: List[List['cirq.Operation']],
+        max_search_radius: int,
```

Comment:
Consider renaming the `max_search_radius` argument to `lookahead_radius`, if you think the latter is more descriptive of how we are using this argument. 
```suggestion
        lookahead_radius: int,
```